sklearn/feature_extraction/_dict_vectorizer.py

```diff
@@ -384,6 +389,8 @@ def get_feature_names_out(self, input_features=None):
         feature_names_out : ndarray of str objects
             Transformed feature names.
         """
+        check_is_fitted(self)
```

Comment:
I think `get_feature_names_out` raises `ValueError` if the right attributes are not set:

https://github.com/scikit-learn/scikit-learn/blob/353b05b0777dc4da36825f8071461d1ea53ab5fd/sklearn/utils/validation.py#L1949-L1950

And this I think is the case (or should be) for all estimators.

cc @thomasjpfan

Reply:
To my understanding, the attribute `self.feature_names_` which is accessed here in `get_feature_names_out()`  two lines below in line 394 will only exist after calling `fit()`. 

Calling `get_feature_names_out()` before `fit()` is currently raising: `AttributeError: 'DictVectorizer' object has no attribute 'feature_names_'` when executing this code example:
```python
from sklearn.feature_extraction import DictVectorizer

dv = DictVectorizer()
dv.get_feature_names_out()
```

This PR proposes to raise a `NotFittedError` instead. Running the snippet from above on the code of the PR results in:
`NotFittedError: This DictVectorizer instance is not fitted yet. Call 'fit' with appropriate arguments before using this estimator.`

I guess it's a matter of taste, which option is the more favorable API. I've created this PR, based on @glemaitre's sentiment in #24816 that an exception of type `NotFittedError` would be favorable.

What is your impression so far?

Reply:
I was looking as well at this matter. It seems that the transformer that uses `ClassNamePrefixFeaturesOutMixin` will raise a `NotFittedError` because we check `check_is_fitted(self, "_n_features_out").

`NotFittedError` seems to be a more informative error than stating that `n_features_in_` is missing. I would rather make the `check_is_fitted` everywhere.

Reply:
That works for me.